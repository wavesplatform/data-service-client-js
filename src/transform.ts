import { Asset, IAssetJSON, BigNumber } from '@waves/data-entities';
import { ApiTypes, TCandleBase, TApiResponse, TApiElement, TransformationResult } from './types';
import { id } from './utils';

const transformer = (response: TApiResponse): TransformationResult | TransformationResult[] => {
  switch (response.__type) {
    case ApiTypes.List:
      return response.data.map(transformSingleElement);
    default:
      return transformSingleElement(response);
  }
};

const transformSingleElement = (el: TApiElement): TransformationResult => {
  if (el === null) return null;
  else {
    switch (el.__type) {
      case ApiTypes.Asset:
        return transformAsset(el.data);
      case ApiTypes.Alias:
        return el.data;
      case ApiTypes.Pair:
        return transformPair(el.data);
      case ApiTypes.Candle:
        return transformCandle(el.data);
      case ApiTypes.Transaction:
        return el.data;
      default:
        throw 'Transformation Error';
    }
  }
};

const transformAsset = (data: IAssetJSON | null): Asset | null =>
  data === null ? null : new Asset(data);

const transformPair = id;

const toBigNumber = (v: string | number | null) =>
  v ? new BigNumber(v) : null;

export const transformCandle = (
  candle: TCandleBase<string | number>
): TCandleBase<BigNumber | null> => ({
  ...candle,
  open: toBigNumber(candle.open),
  close: toBigNumber(candle.close),
  high: toBigNumber(candle.high),
  low: toBigNumber(candle.low),
  weightedAveragePrice: toBigNumber(candle.weightedAveragePrice),
  volume: toBigNumber(candle.volume),
  quoteVolume: toBigNumber(candle.quoteVolume),
});

export default transformer;
