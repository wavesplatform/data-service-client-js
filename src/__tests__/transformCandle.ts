import { transformCandle } from '../transform';
import BigNumber from 'bignumber.js';

describe('Candles transformation', () => {
  it('should transform null candle', () => {
    const time = new Date();
    const emptyCandle = {
      time,
      open: null,
      close: null,
      high: null,
      low: null,
      volume: null,
      quoteVolume: null,
      weightedAveragePrice: null,
      maxHeight: null,
      txsCoung: null,
    };
    expect(transformCandle(emptyCandle)).toEqual(emptyCandle);
  });

  it('should transform candle with number', () => {
    const time = new Date();
    const candle = {
      time,
      open: 0.1,
      close: 0.9,
      high: 1,
      low: 0.001,
      volume: 100.2,
      quoteVolume: 0.001,
      weightedAveragePrice: 0.112,
      maxHeight: 1832,
      txsCoung: 12,
    };
    const newCandle = transformCandle(candle);
    expect(newCandle.open).toBeInstanceOf(BigNumber);
    expect(newCandle.close).toBeInstanceOf(BigNumber);
    expect(newCandle.high).toBeInstanceOf(BigNumber);
    expect(newCandle.low).toBeInstanceOf(BigNumber);
    expect(newCandle.volume).toBeInstanceOf(BigNumber);
    expect(newCandle.quoteVolume).toBeInstanceOf(BigNumber);
    expect(typeof newCandle.maxHeight).toEqual('number');
    expect(typeof newCandle.txsCoung).toEqual('number');
  });

  it('should transform candle with string', () => {
    const time = new Date();
    const candle = {
      time,
      open: '0.1',
      close: '0.9',
      high: '1',
      low: '0.001',
      volume: '100.2',
      quoteVolume: '0.001',
      weightedAveragePrice: '0.112',
      maxHeight: 1832,
      txsCoung: 12,
    };
    const newCandle = transformCandle(candle);
    expect(newCandle.open).toBeInstanceOf(BigNumber);
    expect(newCandle.close).toBeInstanceOf(BigNumber);
    expect(newCandle.high).toBeInstanceOf(BigNumber);
    expect(newCandle.low).toBeInstanceOf(BigNumber);
    expect(newCandle.volume).toBeInstanceOf(BigNumber);
    expect(newCandle.quoteVolume).toBeInstanceOf(BigNumber);
    expect(typeof newCandle.maxHeight).toEqual('number');
    expect(typeof newCandle.txsCoung).toEqual('number');
  });

  it('should transform correct', () => {
    const time = new Date(1544800500076);
    const candle = {
      time,
      open: '0.1',
      close: 0.9,
      high: '1',
      low: '0.001',
      volume: 100.2,
      quoteVolume: '0.001',
      weightedAveragePrice: 0.11,
      maxHeight: 1832,
      txsCoung: 12,
    };
    const newCandle = transformCandle(candle);
    expect(newCandle).toMatchSnapshot();
  });
});
