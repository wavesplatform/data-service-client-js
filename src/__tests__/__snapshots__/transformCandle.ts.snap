// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Candles transformation should transform correct 1`] = `
Object {
  "close": "0.9",
  "high": "1",
  "low": "0.001",
  "maxHeight": 1832,
  "open": "0.1",
  "quoteVolume": "0.001",
  "time": 2018-12-14T15:15:00.076Z,
  "txsCoung": 12,
  "volume": "100.2",
  "weightedAveragePrice": "0.11",
}
`;
